spatialos-sdk/src/worker/entity.rs

```diff
@@ -36,36 +45,151 @@ impl Entity {
             user_handle: data_ptr as *mut _,
         };
 
+        self.components.insert(C::ID, raw_data);
+
+        Ok(())
+    }
+
+    pub(crate) fn add_raw(&mut self, component: &Worker_ComponentData) -> Result<(), String> {
+        let id = component.component_id;
+
+        self.pre_add_check(id)?;
+
+        // Call copy on the component data. We don't own this Worker_ComponentData.
+        let vtable = self.database.get_vtable(id).unwrap();
+        let copy_data_func = vtable
+            .component_data_copy
+            .unwrap_or_else(|| panic!("No component_data_free method defined for {}", id));
```

Comment:
It looks like you need to unwrap here because all of the fields on `Worker_ComponentVtable` are optional (it's a bit unclear to me why the C bindings are generated this way but 🤷). In practice, though does this make sense? Is it ever valid for any of these vtable functions to not be provided? I think the implementation of the `Component` trait currently guarantees that all the components would be supplied. Even beyond that, does the C SDK allow vtable function pointers to be null? I would expect that to be an error.

If it's the case that all fields will always be present, then it would be good to use a different `VTable` type that has the fields as non-optional, statically avoiding the need for this `unwrap`.

Reply:
The `Option<func>` syntax is generated by `bindgen` and my guess is that it signifies that you can pass a `null` ptr down to the C API, which is valid behaviour.

For example, if you want schema objects to be passed back up instead of letting the C API do the deserialization, you give it a vtable full of null-ptrs as the default.

In our implementation, this is never the case - but the underlying C object is like that for a reason. That doesn't mean we can't force `VTable` instances to fully declare it, particularly if we plan to ring fence the default generated code and all associated with it behind a feature flag.

Reply:
Does the C API support partially-null vtables, though? It seems like vtables either have to be fully specified or fully `null`. Is it ever meaningful/valid for only some of the function pointers in the vtable to be null?

Reply:
I'm not sure. cc @davedissian ?

Reply:
Yes, it does fully support it (although in general, not sure what the use cases would be). Perhaps you want to have specialized serialization functions for components, but commands go via a general reflection based serializer (?)

Reply:
Interesting! In that case, it makes sense to keep the C type definitions the same and make an internal vtable type where the fields aren't optional to better represent that all of the functions will always be present for the Rust types.